File: AliasRenderer.m
/*
Copyright (C) 2010-2012 Kristian Duske

This file is part of TrenchBroom.

TrenchBroom is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

TrenchBroom is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
*/

#import "AliasRenderer.h"
#import "Alias.h"
#import "AliasSkin.h"
#import "AliasFrame.h"
#import "Texture.h"
#import "Vbo.h"
#import "Math.h"
#import "Entity.h"

@implementation AliasRenderer

- (id)initWithAlias:(Alias *)theAlias skinIndex:(int)theSkinIndex vbo:(Vbo *)theVbo palette:(NSData *)thePalette {
    NSAssert(theAlias != nil, @"alias must not be nil");
    NSAssert(theSkinIndex >= 0, @"skin index must be at least 0");
    NSAssert(theVbo != NULL, @"VBO must not be nil");
    NSAssert(thePalette != nil, @"palette must not be nil");
    
    if ((self = [self init])) {
        alias = [theAlias retain];
        skinIndex = theSkinIndex;
        vbo = theVbo;
        vboBlock = NULL;
        palette = [thePalette retain];
    }
    
    return self;
}

- (void)dealloc {
    [alias release];
    if (vboBlock != NULL)
        freeVboBlock(vboBlock);
    [texture release];
    [palette release];
    [super dealloc];
}

- (void)renderWithEntity:(id<Entity>)theEntity {
    [self renderAtOrigin:[theEntity origin] angle:[theEntity angle]];
}

- (void)renderAtOrigin:(const TVector3f *)theOrigin angle:(NSNumber *)theAngle {
    if (vboBlock == nil) {
        AliasSkin* skin = [alias skinWithIndex:skinIndex];
        texture = [[Texture alloc] initWithName:[alias name] skin:skin index:0 palette:palette];
        
        AliasFrame* frame = [alias firstFrame];
        triangleCount = [frame triangleCount];
        int vertexSize = 3 * 8;
        
        vboBlock = allocVboBlock(vbo, triangleCount * vertexSize * sizeof(float));
        mapVbo(vbo);
        
        int address = vboBlock->address;
        uint8_t* vboBuffer = vbo->buffer;
        
        for (int i = 0; i < triangleCount; i++) {
            const TFrameTriangle* triangle = [frame triangleAtIndex:i];
            for (int j = 0; j < 3; j++) {
                const TFrameVertex* vertex = &triangle->vertices[j];
                // GL_T2F_N3F_V3F format
                address = writeVector2f(&vertex->texCoords, vboBuffer, address);
                address = writeVector3f(&vertex->norm, vboBuffer, address);
                address = writeVector3f(&vertex->position, vboBuffer, address);
            }
        }
        
        unmapVbo(vbo);
    }
    
    glTranslatef(theOrigin->x, theOrigin->y, theOrigin->z);
    
    if (theAngle != nil) {
        int intAngle = [theAngle intValue];
        if (intAngle == -1)
            glRotatef(90, 1, 0, 0);
        else if (intAngle == 1)
            glRotatef(-90, 1, 0, 0);
        else
            glRotatef(-intAngle, 0, 0, 1);
    }
    
    glEnable(GL_TEXTURE_2D);
    glPolygonMode(GL_FRONT, GL_FILL);
    glTexEnvf(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_REPLACE);
    [texture activate];
    
    glInterleavedArrays(GL_T2F_N3F_V3F, 0, (const GLvoid *)(long)vboBlock->address);
    glDrawArrays(GL_TRIANGLES, 0, triangleCount * 3);
    
    [texture deactivate];
}

- (const TVector3f *)center {
    return [[alias firstFrame] center];
}

- (const TBoundingBox *)bounds {
    return [[alias firstFrame] bounds];
}

- (const TBoundingBox *)maxBounds {
    return [[alias firstFrame] maxBounds];
}

@end

File: Core/Source/Renderer/AliasRenderer.cpp
/*
 Copyright (C) 2010-2012 Kristian Duske

 This file is part of TrenchBroom.

 TrenchBroom is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 TrenchBroom is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with TrenchBroom.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "AliasRenderer.h"
#include "Model/Assets/Alias.h"
#include "Model/Assets/Palette.h"
#include "Model/Assets/Texture.h"
#include "Model/Map/Entity.h"
#include "Model/Preferences.h"
#include "Renderer/MapRenderer.h"
#include "Renderer/RenderContext.h"
#include "Renderer/Vbo.h"
#include "GL/GLee.h"

namespace TrenchBroom {
    namespace Renderer {
        AliasRenderer::AliasRenderer(Model::Assets::Alias& alias, int skinIndex, Vbo& vbo, Model::Assets::Palette& palette) : m_alias(alias), m_skinIndex(skinIndex), m_vbo(vbo), m_palette(palette), m_vboBlock(NULL), m_texture(NULL) {}

        AliasRenderer::~AliasRenderer() {
            if (m_vboBlock != NULL)
                m_vboBlock->freeBlock();
            if (m_texture != NULL)
                delete m_texture;
        }

        void AliasRenderer::render(Model::Entity& entity) {
            render(entity.origin(), static_cast<float>(entity.angle()));
        }

        void AliasRenderer::render(const Vec3f& position, float angle, float scale) {
            if (m_vboBlock == NULL) {
                Model::Assets::AliasSkin& skin = *m_alias.skins[m_skinIndex];
                m_texture = new Model::Assets::Texture(m_alias.name, skin, 0, m_palette);

                Model::Assets::AliasSingleFrame& frame = m_alias.firstFrame();
                m_triangleCount = static_cast<int>(frame.triangles.size());
                int vertexSize = 3 * 8;

                m_vboBlock = &m_vbo.allocBlock(m_triangleCount * vertexSize * sizeof(float));
                m_vbo.map();
                int offset = 0;
                for (unsigned int i = 0; i < m_triangleCount; i++) {
                    Model::Assets::AliasFrameTriangle& triangle = *frame.triangles[i];
                    for (unsigned int j = 0; j < 3; j++) {
                        Model::Assets::AliasFrameVertex& vertex = triangle.vertices[j];
                        // GL_T2F_N3F_V3F format
                        offset = m_vboBlock->writeVec(vertex.texCoords, offset);
                        offset = m_vboBlock->writeVec(vertex.normal, offset);
                        offset = m_vboBlock->writeVec(vertex.position, offset);
                    }
                }
                m_vbo.unmap();
            }

            glTranslatef(position.x, position.y, position.z);

            if (scale != 1.0f)
                glScalef(scale, scale, scale);

            if (angle != 0.0f) {
                if (angle == -1.0f) glRotatef(90, 1, 0, 0);
                else if (angle == -2.0f) glRotatef(-90, 1, 0, 0);
                else glRotatef(-angle, 0, 0, 1);
            }

            m_texture->activate();
            
            glPushClientAttrib(GL_CLIENT_VERTEX_ARRAY_BIT);
            glInterleavedArrays(GL_T2F_N3F_V3F, 0, (const GLvoid *)(long)m_vboBlock->address);
            glDrawArrays(GL_TRIANGLES, 0, m_triangleCount * 3);
            glPopClientAttrib();

            m_texture->deactivate();
        }

        const Vec3f& AliasRenderer::center() {
            return m_alias.firstFrame().center;
        }

        const BBox& AliasRenderer::bounds() {
            return m_alias.firstFrame().bounds;
        }

        const BBox& AliasRenderer::maxBounds() {
            return m_alias.firstFrame().maxBounds;
        }
    }
}
